Realm/Realm/DatabaseTypes/RealmCollectionBase.cs

```diff
@@ -624,6 +588,100 @@ internal interface IRealmCollectionBase<THandle> : IMetadataObject
         THandle NativeHandle { get; }
     }
 
+    [SuppressMessage("StyleCop.CSharp.MaintainabilityRules", "SA1402:File may only contain a single type", Justification = "NotificationCallbacks are tighly coupled with the collection and it's easier to reason about when they're in the same file.")]
+    internal class NotificationCallbacks<T>
+    {
+        private readonly RealmCollectionBase<T> _parent;
+
+        private readonly Dictionary<KeyPath, (NotificationTokenHandle Token, bool DeliveredInitialNotification, List<NotificationCallbackDelegate<T>> Callbacks)> _subscriptions = new();
+
+        public NotificationCallbacks(RealmCollectionBase<T> parent)
+        {
+            _parent = parent;
+        }
+
+        // Shallow here and everywhere else is a bit of a shortcut we're taking until we have proper keypath filtering.
+        // Once we do, we probably want this to be some keypath identifier that we can pass between managed and native.
+        public void Add(NotificationCallbackDelegate<T> callback, bool shallow)
+        {
+            var keyPath = shallow ? KeyPath.Empty : KeyPath.Full;
+            if (_subscriptions.TryGetValue(keyPath, out var subscription))
+            {
+                if (subscription.DeliveredInitialNotification)
+                {
+                    callback(_parent, null, null);
```

Comment:
Unfortunately from `Realm Database emits an initial notification after retrieving the collection.` in our [docs](https://www.mongodb.com/docs/realm/sdk/dotnet/react-to-changes/#register-a-collection-change-listener) it was not clear to me that "retrieving the collection" = "right when subscribing for notification".
To be honest, I am not sure what "retrieving the collection" would really mean:
1. first time you reference the list?
2. first time you create an instance of the class holding that list
3. etc

But thanks for explaining.

Reply:
Typically, it means when the collection is materialized. There are several conditions where this could happen:
1. When you start iterating - this will block the thread and fetch the collection.
2. When you subscribe for notifications - the collection will be created in the background and you'll get the initial notification.

This is mostly relevant for queries which may be expensive to compute if you have many objects to match against. With the second approach, you construct the query, subscribe for notifications, and when it's ready, you'll get a callback with the initial collection that you can use to wire up the UI.